lib/p2p/Peer.ts

```diff
@@ -271,16 +279,41 @@ class Peer extends EventEmitter {
         resolve();
       };
 
-      const onTimeout = () => {
+      const onError = (err: Error) => {
         cleanup();
-        reject(new Error('Connection timed out.'));
-      };
 
-      this.socket!.once('connect', onConnect);
+        if (!retry) {
+          this.close();
+          reject(err);
+          return;
+        }
+
+        if (Date.now() - startTime + retryDelay > Peer.CONNECTION_RETRIES_MAX_PERIOD) {
+          this.close();
+          reject(errors.CONNECTING_RETRIES_MAX_PERIOD_EXCEEDED);
+          return;
+        }
 
-      this.socket!.once('error', onError);
+        this.logger.debug(
+          `Connection attempt #${retries + 1} to peer (${addressUtils.toString(this.socketAddress)}) ` +
+          `failed: ${err.message}. retrying in ${retryDelay / 1000} sec...`,
+        );
+
+        setTimeout(() => {
+          retryDelay = Math.min(Peer.CONNECTION_RETRIES_MAX_DELAY, retryDelay * 2);
+          retries = retries + 1;
+          this.socket!.connect(address);
+          bind();
+        }, retryDelay);
+      };
 
-      this.connectTimeout = setTimeout(onTimeout, Peer.CONNECTION_TIMEOUT);
+      const bind = () => {
```

Comment:
This would be a bit easier to read if `bind` were declared before `onError` which uses the `bind` method, so just move this block of code up a bit.